src/auth/ui/vue/authForms/manageIdentityCenter.vue

```diff
@@ -0,0 +1,233 @@
+            this.state.setValue(key, value)
+
+            if (key === IdentityCenter.startUrl) {
+                this.errors.startUrl = await this.state.getStartUrlError()
+            }
+
+            this.canSubmit = await this.state.canSubmit()
+        },
+    },
+    watch: {
+        'data.startUrl'(value: string) {
+            this.updateData(IdentityCenter.startUrl, value)
+        },
+        'data.region'(value: string) {
+            this.updateData(IdentityCenter.region, value)
+        },
+    },
+})
+
+const IdentityCenter = {
+    startUrl: 'startUrl',
+    region: 'region',
+} as const
+type IdentityCenterKey = (typeof IdentityCenter)[keyof typeof IdentityCenter]
+type IdentityCenterData = { [IdentityCenter.startUrl]: string; [IdentityCenter.region]: Region['id'] }
+
+/**
+ * Manages the state of Builder ID.
+ */
+abstract class BaseIdentityCenterState implements AuthStatus {
+    protected _data: Partial<IdentityCenterData>
+    protected _stage: IdentityCenterStage = stages.START
+
+    constructor() {
+        this._data = {
+            startUrl: '',
+            region: '',
+        }
+    }
+
+    abstract get id(): AuthFormId
+    protected abstract _startIdentityCenterSetup(): Promise<void>
+    abstract isAuthConnected(): Promise<boolean>
+
+    setValue(key: IdentityCenterKey, value: string) {
+        this._data[key] = value
+    }
+
+    getValue(key: IdentityCenterKey) {
+        return this._data[key]
+    }
+
+    async startIdentityCenterSetup(): Promise<void> {
+        this._stage = stages.WAITING_ON_USER
+        return this._startIdentityCenterSetup()
+    }
+
+    async stage(): Promise<IdentityCenterStage> {
+        const isAuthConnected = await this.isAuthConnected()
+        this._stage = isAuthConnected ? stages.CONNECTED : stages.START
+        return this._stage
+    }
+
+    async signout(): Promise<void> {
+        return client.signoutIdentityCenter()
+    }
+
+    async getRegion(): Promise<Region> {
+        return client.getIdentityCenterRegion()
+    }
+
+    async getStartUrlError() {
+        const error = await client.getSsoUrlError(this._data.startUrl)
+        return error ?? ''
+    }
+
+    async canSubmit() {
+        const allFieldsFilled = Object.values(this._data).every(val => !!val)
+        const hasErrors = await this.getStartUrlError()
+        return allFieldsFilled && !hasErrors
+    }
+}
+
+export class CodeWhispererIdentityCenterState extends BaseIdentityCenterState {
+    override get id(): AuthFormId {
+        return authForms.IDENTITY_CENTER_CODE_WHISPERER
+    }
+
+    protected override async _startIdentityCenterSetup(): Promise<void> {
+        const data = await this.getSubmittableDataOrThrow()
+        return client.startIdentityCenterSetup(data.startUrl, data.region)
+    }
+
+    private async getSubmittableDataOrThrow(): Promise<IdentityCenterData> {
+        return this._data as IdentityCenterData
+    }
```

Comment:
Feels clunky. The only difference between the concrete class and the abstract class is `isCodeWhispererIdentityCenterConnected` + an ID? Why not pass the id to the extension host and call the correct function there?